### PR TITLE
Use internal IP to identify etcd members

### LINKED ISSUE #14: Infrastructure controller failing to fetch etcd member
#### What is the current behavior?
Infrastructure is throwing errors trying to retrieve etcd members by name

#### What is the expected behavior?
It should successfully retrieve node member

### Evidence
```
2019-03-07T15:36:54.898Z        INFO    Found etcd member named "ip-172-20-39-46.ec2.internal" with no corresponding Kubernetes node
2019-03-07T15:36:54.899Z        DEBUG   Getting all etcd node pools
2019-03-07T15:36:54.962Z        DEBUG   Checking for existence of node "ip-172-20-39-46.ec2.internal" in etcd node pool "8729589c-6262-4466-a474-7b37e2ffd13f"
2019-03-07T15:36:54.981Z        INFO    Dropping node "fb1b4ff2-71a1-4036-891d-20863da3d246" out of the queue: checking if node "ip-172-20-39-46.ec2.internal" exists in cloud: attempting to get node "ip-172-20-39-46.ec2.internal" from pool "8729589c-6262-4466-a474-7b37e2ffd13f": server responded with status 400: {"message":"Invalid params: \"node_id\" must be
a valid GUID."}
2019-03-07T15:36:54.981Z        ERROR   checking if node "ip-172-20-39-46.ec2.internal" exists in cloud: attempting to get node "ip-172-20-39-46.ec2.internal" from pool "8729589c-6262-4466-a474-7b37e2ffd13f": server responded with status 400: {"message":"Invalid params: \"node_id\" must be a valid GUID."}
github.com/containership/infrastructure-controller/vendor/github.com/containership/cluster-manager/pkg/log.Error
        /go/src/github.com/containership/infrastructure-controller/vendor/github.com/containership/cluster-manager/pkg/log/log.go:94
github.com/containership/infrastructure-controller/pkg/controller.(*EtcdRemovalController).processNextWorkItem
        /go/src/github.com/containership/infrastructure-controller/pkg/controller/etcd_removal.go:141
github.com/containership/infrastructure-controller/pkg/controller.(*EtcdRemovalController).runWorker
        /go/src/github.com/containership/infrastructure-controller/pkg/controller/etcd_removal.go:110
github.com/containership/infrastructure-controller/pkg/controller.(*EtcdRemovalController).(github.com/containership/infrastructure-controller/pkg/controller.runWorker)-fm
        /go/src/github.com/containership/infrastructure-controller/pkg/controller/etcd_removal.go:98
github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
        /go/src/github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/src/github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until
        /go/src/github.com/containership/infrastructure-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
```

### Environment
AWS CKE 1.13.4 Cluster

### LINKED ISSUE #15: Erroneous etcd member name log when no backing instance is available
#### What is the current behavior?
When a new master node is being provisioned, its etcd member representation is added before the actual Kubernetes node is created. Infrastructure controller prints an erroneous log until the backing VM is available.

#### What is the expected behavior?
It should either not log at this point, or print the actual etcd members name.

### Evidence
```
2019-03-07T15:49:59.971Z        INFO    Found etcd member named "" with no corresponding Kubernetes node
```

### Environment
AWS CKE 1.12.x

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

Use internal IP to identify etcd members

This is required because CKE now delegates the responsibility of naming
etcd members to kubeadm, so we can't rely on the member names being the
same as the Containership node ID.

See https://github.com/containership/infrastructure-controller/issues/14.

#### What issue(s) does this fix?
* Fixes #14
* Fixes #15

#### Setup

- Provision HA cluster with a bunch of masters
- Build and push image, update coordinator to not sync cluster-management plugin, update image on infra-controller deployment

#### Instructions

- Scale down master pool
- See etcd members properly removed

```
2019-03-08T18:25:14.665Z        INFO    Starting Containership Infrastructure Controller...
2019-03-08T18:25:14.665Z        INFO    Version: v2.0.0-4-g6797fa7 (6797fa7) built @ 2019-03-08 18:21:55 +0000 UTC
2019-03-08T18:25:14.665Z        INFO    Go Version: go1.10.1
2019-03-08T18:25:14.665Z        DEBUG   (env.environment) {
 organizationID: (string) (len=36) "5813c2b6-9bf2-43ba-bc60-f3711e4e4d8a",
 clusterID: (string) (len=36) "23023420-0915-4210-bc31-c00a605810ed",
 clusterToken: (string) (len=324) <redacted>,
 etcdEndpoint: (string) (len=104) "etcd.23023420-0915-4210-bc31-c00a605810ed.5813c2b6-9bf2-43ba-bc60-f3711e4e4d8a.stage.cloud.cship.io:2379",
 csCloudEnvironment: (string) (len=5) "stage",
 provisionBaseURL: (string) (len=40) "https://stage-provision.containership.io"
}

2019-03-08T18:25:14.665Z        INFO    Using in cluster k8s config
2019-03-08T18:25:14.666Z        INFO    Setting up event handlers
2019-03-08T18:25:14.666Z        INFO    Starting controller
2019-03-08T18:25:14.766Z        INFO    Starting workers
2019-03-08T18:25:14.766Z        INFO    Started workers
2019-03-08T18:25:37.825Z        DEBUG   etcd member "354440f8-9898-49bc-8071-51f2b0f31aef" assumed unhealthy because we can't connect (error: etcd client connect to endpoints [https://10.136.142.224:2380] failed: context deadline exceeded)
2019-03-08T18:25:37.825Z        INFO    Requesting etcd member remove for member with name "354440f8-9898-49bc-8071-51f2b0f31aef"
2019-03-08T18:25:37.848Z        INFO    etcd member "354440f8-9898-49bc-8071-51f2b0f31aef" (ID 5b226789eb833535) removed successfully
2019-03-08T18:28:41.522Z        DEBUG   etcd member "2345d53d-34f4-49fd-b6ab-31e748d17cab" assumed unhealthy because we can't connect (error: etcd client connect to endpoints [https://10.136.137.159:2380] failed: context deadline exceeded)
2019-03-08T18:28:41.522Z        INFO    Requesting etcd member remove for member with name "2345d53d-34f4-49fd-b6ab-31e748d17cab"
2019-03-08T18:28:41.539Z        INFO    etcd member "2345d53d-34f4-49fd-b6ab-31e748d17cab" (ID 8f8c112fbe6ae8ca) removed successfully
2019-03-08T18:30:58.825Z        DEBUG   etcd member "bb1e10b6-21bc-4c19-b478-bcc414e4b14e" assumed unhealthy because we can't connect (error: etcd client connect to endpoints [https://10.136.140.107:2380] failed: context deadline exceeded)
2019-03-08T18:30:58.825Z        INFO    Requesting etcd member remove for member with name "bb1e10b6-21bc-4c19-b478-bcc414e4b14e"
2019-03-08T18:30:58.842Z        INFO    etcd member "bb1e10b6-21bc-4c19-b478-bcc414e4b14e" (ID eed7ef8e5ea2a98c) removed successfully
```

### Dependencies
* None

### CODE REVIEW

File: pkg/controller/etcd_removal.go [hunk 1]
```diff
@@ -174,58 +178,92 @@ func (c *EtcdRemovalController) enqueueNode(obj interface{}) {
...
+	if err != nil {
+		return errors.Wrap(err, "listing nodes running etcd")
+	}
+
+	var candidate *etcdserverpb.Member
+
+	for _, member := range members {
+		foundInCloud := false
+
+	LoopOverAddresses:
```
Comment (mattkelly) [inline pkg/controller/etcd_removal.go]: Pretty gnarly nested for loops here. I'll consider adding a helper function.

**Comment by mattkelly** [comment]
Marked as `needs-retested` because I'd like to run a test on AWS just to sanity check. There should be nothing at all that's provider-dependent, though.

File: pkg/controller/etcd_removal.go [hunk 2]
```diff
@@ -174,58 +178,92 @@ func (c *EtcdRemovalController) enqueueNode(obj interface{}) {
 	c.workqueue.AddRateLimited(key)
 }
 
-// syncHandler surveys the state of the system and determines if etcd has any
+// syncHandler surveys the state of the system and etermines if etcd has any
```
Comment (mattkelly) [inline pkg/controller/etcd_removal.go]: et tu, Brute?

**Comment by mattkelly** [comment]
Looks good on AWS 🚢

**Comment by mattkelly** [comment]
Marked as WIP while I refactor some for loops and write tests. CI is flaky right now.

**Comment by mattkelly** [comment]
Updated to simplify logic and add some tests. This is good for a review.

File: pkg/controller/etcd_removal.go [hunk 3]
```diff
@@ -180,52 +184,93 @@ func (c *EtcdRemovalController) enqueueNode(obj interface{}) {
...
+		return false
+	}
+
+	for _, addr := range member.PeerURLs {
+		// Ignore errors - an etcd member wouldn't be able to come up with
+		// a malformed peer URL
+		u, _ := url.Parse(addr)
+
+		for _, node := range nodes {
+			if node.Addresses != nil && node.Addresses.InternalIP == u.Hostname() {
```
Comment (ashleyschuett) [inline pkg/controller/etcd_removal.go]: This is my only concern. Was this checked across multiple providers such as aws / gcp ? I thought aws set this to their weird private dns?

Comment (mattkelly) [reply pkg/controller/etcd_removal.go]: This is a Containership Cloud node, not a Kubernetes node :) @mikestoltz added the IPs to the node response recently.

**Pull request merged**